front50-core/src/main/java/com/netflix/spinnaker/front50/model/plugins/PluginInfoService.java

```diff
@@ -62,9 +69,48 @@ public PluginInfo findById(@Nonnull String pluginId) {
     return repository.findById(pluginId);
   }
 
-  public PluginInfo upsert(@Nonnull PluginInfo pluginInfo) {
+  /** Upserts the PluginInfo to the repository and sends Plugin Events to Echo */
+  private PluginInfo savePluginInfo(PluginInfo pluginInfo) {
     validate(pluginInfo);
 
+    PluginInfo oldPluginInfo;
+    try {
+      oldPluginInfo = findById(pluginInfo.getId());
+    } catch (NotFoundException e) {
+      oldPluginInfo = null;
+    }
+
+    PluginInfoDelta delta = new PluginInfoDelta(pluginInfo, oldPluginInfo);
+
+    delta.addedReleases.forEach(
+        release -> postEvent(PluginEventType.PUBLISHED, pluginInfo, release));
+
+    String newVersion =
+        Optional.ofNullable(delta.newPreferredRelease)
+            .map(PluginInfo.Release::getVersion)
+            .orElse(null);
+
+    String oldVersion =
+        Optional.ofNullable(delta.oldPreferredRelease)
+            .map(PluginInfo.Release::getVersion)
+            .orElse(null);
+
+    if ((newVersion == null && oldVersion != null
+        || (newVersion != null && !newVersion.equals(oldVersion)))) {
+      postEvent(PluginEventType.PREFERRED_VERSION_UPDATED, pluginInfo, delta.newPreferredRelease);
+    }
+
+    if (oldPluginInfo == null) {
+      repository.create(pluginInfo.getId(), pluginInfo);
+    } else {
+      repository.update(pluginInfo.getId(), pluginInfo);
+    }
+
+    return pluginInfo;
```

Comment:
I think the events should be posted after the plugin is created or updated.   That will be better from both a consistency and error handling perspective.

Reply:
agreed, fixed